[JSC] Add WebAssembly.JSTag

Constellation opened the pull request:
#### 6bf8f0f20ef1ee92e5ba53c865c339a1253a080f
<pre>
[JSC] Add WebAssembly.JSTag
<a href="https://bugs.webkit.org/show_bug.cgi?id=286103">https://bugs.webkit.org/show_bug.cgi?id=286103</a>
<a href="https://rdar.apple.com/136036182">rdar://136036182</a>

Reviewed by Keith Miller.

This patch adds WebAssembly.JSTag for modern wasm exception handling.
This is a special tag which can catch random JS-thrown values in wasm
catch clauses. When catching non JSWebAssemblyException via JSTag,
we wrap the value with JSWebAssemblyException and pass it to wasm catch
clause. Also when rethrowing an exception, we unwrap JSWebAssemblyException
for JSTag.

* JSTests/wasm/assert.js:
* JSTests/wasm/stress/jstag.js: Added.
(import.as.assert.from.string_appeared_here.throw_ref):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::UnwindFunctor::UnwindFunctor):
(JSC::UnwindFunctor::operator() const):
(JSC::UnwindFunctor::notifyDebuggerOfUnwinding):
(JSC::Interpreter::unwind):
* Source/JavaScriptCore/runtime/Exception.cpp:
(JSC::Exception::wrapValueForJSTag):
* Source/JavaScriptCore/runtime/Exception.h:
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::WASM_IPINT_EXTERN_CPP_DECL):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmTag.cpp:
(JSC::Wasm::Tag::jsExceptionTag):
* Source/JavaScriptCore/wasm/WasmTag.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeInformation::signatureForJSException):
(JSC::Wasm::TypeInformation::TypeInformation):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/JavaScriptCore/wasm/js/JSWebAssembly.cpp:
(JSC::JSWebAssembly::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/js/WebAssemblyExceptionConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/289094@main">https://commits.webkit.org/289094@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/178648749166fd9c6c110543a34cd4778e850ea9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85308 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5043 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39739 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90435 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36350 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5132 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13019 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66322 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24132 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88354 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3908 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77489 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46604 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3794 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31750 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/iframe-loading-lazy-multiple-queued-navigations.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35418 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/78276 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74518 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32588 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91929 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/84349 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12655 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9248 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/74898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12884 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73326 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/74015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18390 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16824 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4676 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13302 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12598 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18060 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/106739 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12428 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25729 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15921 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14179 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on JSTests/wasm/stress/jstag.js:
```diff
@@ -0,0 +1,44 @@
+}
+
+let code = new Uint8Array([
+    0x00,0x61,0x73,0x6d,0x01,0x00,0x00,0x00,0x01,0x8f,0x80,0x80,0x80,0x00,0x03,0x60,0x01,0x6f,0x00,0x60,0x01,0x6f,0x01,0x6f,0x60,0x00,0x02,0x6f,0x69,0x02,0x98,0x80,0x80,0x80,0x00,0x02,0x01,0x69,0x03,0x74,0x61,0x67,0x04,0x00,0x00,0x01,0x69,0x09,0x74,0x68,0x72,0x6f,0x77,0x5f,0x72,0x65,0x66,0x00,0x01,0x03,0x84,0x80,0x80,0x80,0x00,0x03,0x01,0x01,0x01,0x07,0xda,0x80,0x80,0x80,0x00,0x03,0x17,0x63,0x61, ...
+]);
```
Can you paste the wat here?

Constellation replied:
Attached the original code from WPT.

kmiller68 commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -728,12 +729,22 @@ class UnwindFunctor {
 #if ENABLE(WEBASSEMBLY)
                 if (m_catchableFromWasm) {
                     auto* wasmCallee = static_cast<Wasm::Callee*>(nativeCallee);
+                    if (wasmCallee->compilationMode() == Wasm::CompilationMode::WasmToJSMode) {
+                        // https://webassembly.github.io/exception-handling/js-api/#create-a-host-function
+                        if (!m_wasmTag)
+                            m_wasmTag = &Wasm::Tag::jsExceptionTag();
+                    }
```
It's probably not super perf important but why not hoist this to the constructor? That would take it out of the loop.

Constellation replied:
This is important since we would like to set this tag only when we see WasmToJSMode.

kmiller68 reviewed (approved):
r=me with comment.

Constellation committed:
[JSC] Add WebAssembly.JSTag
https://bugs.webkit.org/show_bug.cgi?id=286103
rdar://136036182

Reviewed by Keith Miller.

This patch adds WebAssembly.JSTag for modern wasm exception handling.
This is a special tag which can catch random JS-thrown values in wasm
catch clauses. When catching non JSWebAssemblyException via JSTag,
we wrap the value with JSWebAssemblyException and pass it to wasm catch
clause. Also when rethrowing an exception, we unwrap JSWebAssemblyException
for JSTag.

* JSTests/wasm/assert.js:
* JSTests/wasm/stress/jstag.js: Added.
(import.as.assert.from.string_appeared_here.throw_ref):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::UnwindFunctor::UnwindFunctor):
(JSC::UnwindFunctor::operator() const):
(JSC::UnwindFunctor::notifyDebuggerOfUnwinding):
(JSC::Interpreter::unwind):
* Source/JavaScriptCore/runtime/Exception.cpp:
(JSC::Exception::wrapValueForJSTag):
* Source/JavaScriptCore/runtime/Exception.h:
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::WASM_IPINT_EXTERN_CPP_DECL):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmSlowPaths.cpp:
(JSC::LLInt::WASM_SLOW_PATH_DECL):
* Source/JavaScriptCore/wasm/WasmTag.cpp:
(JSC::Wasm::Tag::jsExceptionTag):
* Source/JavaScriptCore/wasm/WasmTag.h:
* Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp:
(JSC::Wasm::TypeInformation::signatureForJSException):
(JSC::Wasm::TypeInformation::TypeInformation):
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
* Source/JavaScriptCore/wasm/js/JSWebAssembly.cpp:
(JSC::JSWebAssembly::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/js/WebAssemblyExceptionConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/289094@main
